File: Striver/DP/Q49.java
package Striver.DP;

public class Q49{
    // matrix chain multiplication

    // tabulation solution - Time: O(n^3), Space: O(n^2)

    public static int matrixChainMultiplication(int[] arr){
        int n = arr.length;
        int[][] dp = new int[n][n];

        for(int i=1; i<n; i++)
            dp[i][i] = 0;

        for(int i = n-1; i>=1; i--){
            for(int j = i+1; j<n; j++){
                dp[i][j] = Integer.MAX_VALUE;
                for(int k = i; k<j; k++){
                    int temp = dp[i][k] + dp[k+1][j] + arr[i]*arr[k+1]*arr[j+1];
                    dp[i][j] = Math.min(dp[i][j], temp);
                }
            }
        }
        return dp[1][n-1];
    }

    public static void main(String[] args) {
        int[] arr = {40, 20, 30, 10, 30};
        System.out.println(matrixChainMultiplication(arr));
    }
}

File: Striver/DP/Q49.py
# matrix chain multiplication

# tabulation
def matrix_chain_multiplication(arr):
    n = len(arr)
    dp = [[0 for i in range(n)] for j in range(n)]
    for i in range(1, n):
        dp[i][i] = 0

    for i in range(n-1,0,-1):
        for j in range(i+1,n):
            dp[i][j] = float('inf')
            for k in range(i,j):
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k+1][j] + arr[i-1]*arr[k]*arr[j])
            
    return dp[1][n-1]

arr = [40, 20, 30, 10, 30]
print(matrix_chain_multiplication(arr))